lib/stages/visSharedMemWriter.cpp

```diff
@@ -0,0 +1,295 @@
+
+    frameID frame_id(in_buf);
+
+    // Create the semaphore, and gain first access to it
+    sem = sem_open(
+            sem_name.c_str(),
+            (O_CREAT | O_EXCL),
+            (S_IRUSR | S_IWUSR),
+            1
+        );
+
+    if (sem == SEM_FAILED) {
+        perror("sem_open");
+        exit(errno);
+    }
+
+    INFO("Semaphore created.\n");
+
+    // Acquire semaphore until shared memory is created
+    CHECK(sem_wait(sem));
+
+
+    // Set up the structure of the ring buffer shared memory
+    // Get one frame for reference
+    wait_for_full_frame(in_buf, unique_name.c_str(), frame_id);
+
+    auto frame = visFrameView(in_buf, frame_id);
+
+    // Set up the structure of the access record shared memory
+    size_t access_record_size = sizeof(uint64_t);
+
+    record_addr = assign_memory<uint64_t*>(fname_access_record, ntime * access_record_size, record_addr);
+
+    // Get properties of stream from first frame and datasetManager
+    std::map<uint32_t, uint32_t> freq_id_map;
+    auto& dm = datasetManager::instance();
+    auto sstate_fut = std::async(&datasetManager::dataset_state<stackState>, &dm, frame.dataset_id);
+    auto istate_fut = std::async(&datasetManager::dataset_state<inputState>, &dm, frame.dataset_id);
+    auto pstate_fut = std::async(&datasetManager::dataset_state<prodState>, &dm, frame.dataset_id);
+    auto fstate_fut = std::async(&datasetManager::dataset_state<freqState>, &dm, frame.dataset_id);
+    auto evstate_fut = std::async(&datasetManager::dataset_state<eigenvalueState>, &dm, frame.dataset_id);
+
+    const inputState* istate = istate_fut.get();
+    const prodState* pstate = pstate_fut.get();
+    const stackState* sstate = sstate_fut.get();
+    const freqState* fstate = fstate_fut.get();
+    const eigenvalueState* evstate = evstate_fut.get();
+
+    
+    if (!istate || !pstate || !fstate) {
+        ERROR("Required datasetState not found for dataset ID {}\nThe following required states "
+                "were found:\ninputState - {:p}\nprodState - {:p}\nfreqState - {:p}\n",
+                frame.dataset_id, (void*)istate, (void*)pstate, (void*)fstate);
+        throw std::runtime_error("Could not write to shared memory.");
+    }
+
+    // Build the frequency index
+    uint ind = 0;
+    for (auto& f : fstate->get_freqs())
+        freq_id_map[f.first] = ind++;
+
+    // Count the eigenvalue index
+    size_t num_ev;
+    if (evstate) {
+        num_ev = evstate->get_num_ev();
+    } else {
+        num_ev = 0;
+    }
+
+    // Figure out the ring buffer structure
+    nfreq = fstate->get_freqs().size();
+    size_t ninput = istate->get_inputs().size();
+    size_t nvis = sstate ? sstate->get_num_stack() : pstate->get_prods().size();
+
+    // Set the alignment (in kB)
+    size_t alignment = 4; // Align on page boundaries
+
+    // Calculate the ring buffer stricture
+    auto layout = visFrameView::calculate_buffer_layout(ninput, nvis, num_ev);
+    data_size = layout.first;
```

Comment:
I think this is the only reason you have done all the dM stuff above. You can do this in three other ways:
- Look if up from the buffer. This is not ideal because often the buffer length is just some number I made up that I know if larger than the frame size.
- Make the same `calculate_buffer_layout` call you do above, but fetch the `ninput`, `nvis` and `num_ev` arguments from the incoming data (they're exposed by the `visFrameView`)
- Last, and probably the best option, add a method (or expose a reference) to the data size on `visFrameView` directly. This you could do by adding a public `const size_t& data_size` to `visFrameView` and then in the constructors initializers list adding `data_size(buffer_layout.first)`.